src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala

```diff
@@ -0,0 +1,244 @@
+  }
+}
+
+trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths {
+  val dataSource: SlickDataSource
+  val methodRepoDAO: MethodRepoDAO
+  val googleServicesDAO: GoogleServicesDAO
+  val executionServiceDAO: ExecutionServiceDAO
+  val batchSize: Int
+  val credential: Credential
+  val pollInterval: FiniteDuration
+
+  import dataSource.dataAccess.driver.api._
+
+  //Get a blob of unlaunched workflows, flip their status, and queue them for submission.
+  def getUnlaunchedWorkflowBatch()(implicit executionContext: ExecutionContext): Future[WorkflowSubmissionMessage] = {
+    val unlaunchedWfOptF = dataSource.inTransaction { dataAccess =>
+      //grab a bunch of unsubmitted workflows
+      dataAccess.workflowQuery.findUnsubmittedWorkflows().take(batchSize).result flatMap { wfRecs =>
+        if (wfRecs.isEmpty) {
+          DBIO.successful(wfRecs)
+        } else {
+          //they should also all have the same submission ID
+          val wfsWithASingleSubmission = wfRecs.filter(_.submissionId == wfRecs.head.submissionId)
+          dataAccess.workflowQuery.batchUpdateStatus(wfsWithASingleSubmission, WorkflowStatuses.Launching).map(_ => wfsWithASingleSubmission)
+        }
+      }
+    }
+
+    unlaunchedWfOptF.map {
+      // submit the batch we found
+      case workflowRecs if workflowRecs.nonEmpty => SubmitWorkflowBatch(workflowRecs.map(_.id))
+
+      // if we didn't find any then look again in a little bit
+      case _ => ScheduleNextWorkflowQuery
+
+    } recover {
+      // if we found some but another actor reserved the first look again immediately
+      case t: RawlsConcurrentModificationException => LookForWorkflows
+    }
+  }
+
+  def buildWorkflowOpts(workspace: WorkspaceRecord, submissionId: UUID, user: RawlsUser, token: String, billingProject: RawlsBillingProject) = {
+    ExecutionServiceWorkflowOptions(
+      s"gs://${workspace.bucketName}/${submissionId}",
+      workspace.namespace,
+      user.userEmail.value,
+      token,
+      billingProject.cromwellAuthBucketUrl
+    )
+  }
+
+  def getWdl(methodConfig: MethodConfiguration)(implicit executionContext: ExecutionContext): ReadWriteAction[String] = {
+    withMethod(methodConfig.methodRepoMethod.methodNamespace, methodConfig.methodRepoMethod.methodName, methodConfig.methodRepoMethod.methodVersion, getUserInfo(credential)) { method =>
+      withWdl(method) { wdl =>
+        DBIO.successful(wdl)
+      }
+    }
+  }
+
+  def getWorkflowRecordBatch(workflowIds: Seq[Long], dataAccess: DataAccess)(implicit executionContext: ExecutionContext) = {
+    dataAccess.workflowQuery.findWorkflowByIds(workflowIds).result map { wfRecs =>
+      val submissionId = wfRecs.head.submissionId
+      if( !wfRecs.forall(_.submissionId == submissionId) ) {
+        throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, "Workflow batch has different submissions!"))
```

Comment:
Ultimately `reserveWorkflowBatch` but that is several steps away so... yeah ok let's keep this check.